flexget/plugins/filter/upgrade.py

```diff
@@ -0,0 +1,185 @@
+from flexget.db_schema import Session
+from flexget.event import event
+from flexget.utils import qualities
+
+log = logging.getLogger('upgrade')
+
+Base = db_schema.versioned_base('upgrade', 0)
+
+entry_actions = {
+    'accept': Entry.accept,
+    'reject': Entry.reject,
+    'fail': Entry.fail
+}
+
+
+class EntryUpgrade(Base):
+    __tablename__ = 'upgrade'
+
+    id = Column(Unicode, primary_key=True, index=True)
+    title = Column(Unicode)
+    _quality = Column('quality', String)
+    quality = quality_property('_quality')
+    proper = Column(Boolean)
+    added = Column(DateTime, index=True)
+
+    def __init__(self):
+        self.added = datetime.now()
+
+    def __str__(self):
+        return '<Upgrade(id=%s,added=%s,quality=%s)>' % \
+               (self.id, self.added, self.quality)
+
+
+def group_entries(entries, identified_by):
+    grouped_entries = defaultdict(list)
+
+    # Group by Identifier
+    for entry in entries:
+        identifier = entry.get('id') if identified_by == 'auto' else entry.render(identified_by)
+        if not identifier:
+            log.debug('No identifier found for', entry['title'])
+            continue
+        grouped_entries[identifier.lower()].append(entry)
+
+    return grouped_entries
+
+
+class LazyUpgrade(object):
+
+    schema = {
+        'oneOf': [
+            {'type': 'boolean'},
+            {
+                'type': 'object',
+                'properties': {
+                    'identified_by': {'type': 'string'},
+                    'tracking': {'type': 'boolean'},
+                    'target': {'type': 'string', 'format': 'quality_requirements'},
+                    'on_lower': {'type': 'string', 'enum': ['accept', 'reject', 'fail', 'skip']}
+                },
+                'additionalProperties': False
+            }
+        ]
+    }
+
+    def prepare_config(self, config):
+        if not config or isinstance(config, bool):
+            config = {}
+        config.setdefault('identified_by', 'auto')
+        config.setdefault('tracking', True)
+        config.setdefault('on_lower', 'skip')
+        config.setdefault('target', None)
+        return config
+
+    def on_task_filter(self, task, config):
+        if not config:
+            return
+
+        config = self.prepare_config(config)
+
+        grouped_entries = group_entries(task.entries, config['identified_by'])
+
+        with Session() as session:
+            # Prefetch Data
+            existing_ids = session.query(EntryUpgrade).filter(EntryUpgrade.id.in_(grouped_entries.keys())).all()
+            existing_ids = {e.id: e for e in existing_ids}
+
+            for identifier, entries in grouped_entries.items():
+                if not entries:
+                    continue
+
+                existing = existing_ids.get(identifier)
+                if not existing:
+                    # No existing, skip
+                    continue
+
+                target_downloaded = False
```

Comment:
I'm assuming this will be used somehow later, so this is just a reminder that this variable is currently _never_ `True`.

Reply:
yeah for later use, still wip